internal/pkg/peer/blocksprovider/bft_censorship_monitor.go

```diff
@@ -0,0 +1,277 @@
+	BlockProgress() (uint64, time.Time)
+}
+
+// We ask for a stream of headers from all sources except the one supplying blocks.
+// We track the progress of header receivers against the block reception progress.
+// If there is a header that is ahead of the last block, and a timeout had passed since that header was received, we
+// declare that censorship was detected.
+// When censorship is detected, errCensorship is sent to the errorCh which can be read by ErrorsChannel() method.
+type bftCensorshipMonitor struct {
+	chainID          string
+	headerVerifier   BlockVerifier
+	requester        *DeliveryRequester
+	fetchSources     []*orderers.Endpoint
+	blockSourceIndex int
+
+	progressReporter     BlockProgressReporter // Provides the last block number and time
+	suspicion            bool                  // If a suspicion is pending
+	suspicionTime        time.Time             // The reception time of the header that is ahead of the block receiver
+	suspicionBlockNumber uint64                // The block number of the header that is ahead of the block receiver
+
+	mutex          sync.Mutex
+	stopFlag       bool
+	stopCh         chan struct{}
+	errorCh        chan error
+	hdrRcvMonitors map[string]*headerReceiverMonitor
+
+	logger *flogging.FabricLogger
+}
+
+type headerReceiverMonitor struct {
+	headerReceiver *bftHeaderReceiver
+	failureCounter uint
+	retryDeadline  time.Time
+}
+
+func NewBFTCensorshipMonitor(
+	chainID string,
+	verifier BlockVerifier,
+	requester *DeliveryRequester,
+	fetchSources []*orderers.Endpoint,
+	blockSourceIndex int,
+) *bftCensorshipMonitor {
+	m := &bftCensorshipMonitor{
+		chainID:          chainID,
+		headerVerifier:   verifier,
+		requester:        requester,
+		fetchSources:     fetchSources,
+		stopCh:           make(chan struct{}),
+		errorCh:          make(chan error),
+		hdrRcvMonitors:   make(map[string]*headerReceiverMonitor),
+		blockSourceIndex: blockSourceIndex,
+	}
+
+	return m
+}
+
+// Monitor the progress of headers and compare to the progress of block fetching, trying to detect block censorship.
+// Continuously try and relaunch the goroutines that monitor individual orderers. If an  orderer is faulty we increase
+// the interval between retries but never quit.
+func (m *bftCensorshipMonitor) Monitor() {
+	m.logger.Debug("Starting to monitor block and header fetching progress")
+	for {
+		if err := m.launchHeaderReceivers(); err != nil {
+			m.logger.Warningf("Failure while launching header receiver monitors: %s", err)
+		}
+
+		select {
+		case <-m.stopCh:
+			m.logger.Debug("Stopping to monitor progress")
+			m.errorCh <- &errStopping{message: "received a stop signal"}
+			return
+		case <-time.After(bftBlockCensorshipTimeout / 100):
+			if m.detectBlockCensorship() {
+				m.errorCh <- &errCensorship{message: fmt.Sprintf("block censorship detected, endpoint: %s", m.fetchSources[m.blockSourceIndex])}
+				m.logger.Debug("Block censorship detected, stopping to monitor progress")
+				m.Stop()
```

Comment:
I don't understand. Who relaunches this function if we exit?

Reply:
This method should be run using a dedicated goroutine.